### PR TITLE
ENT-3633 - (bug) Hourly snapshots being produced for non-openshift products

### PR DESCRIPTION
**Pull request by lindseyburnett**
On the develop branch, if you follow the instructions below for seeding data, then running some tallies, RHEL Hourly snapshots were being produced...this first query and table of results below is an example.  I don't have instructions included for putting test data into the insights database for the HBI portion of this.....but you'll need that in order to get RHEL snapshots.

```sql
SELECT DISTINCT product_id,
    --account_number,
    granularity
FROM tally_snapshots
WHERE sla = '_ANY'
    AND usage = '_ANY'
     AND granularity = 'HOURLY'
    --AND account_number = 'account123'
GROUP BY product_id,
    --account_number,
    granularity
ORDER BY --account_number,
    granularity ASC;
```

```

+---------------------------+-----------+
|product_id                 |granularity|
+---------------------------+-----------+
|OpenShift-dedicated-metrics|HOURLY     |
|OpenShift-metrics          |HOURLY     |
|RHEL                       |HOURLY     |
|RHEL Desktop               |HOURLY     |
|RHEL for x86               |HOURLY     |
|RHEL Server                |HOURLY     |
+---------------------------+-----------+

```

# Test instructions
---
## Clear data & insert test event data
```sql
TRUNCATE tally_snapshots CASCADE;
TRUNCATE hosts CASCADE;
TRUNCATE events CASCADE;

INSERT INTO events (id, account_number, timestamp, data, event_type, event_source, instance_id) VALUES ('1edc517a-d049-4361-9804-e4111ce061ae', 'account123', '2021-03-10 19:00:00.000000', '{"role":"ocp","sla": "Premium", "usage": "Development/Test", "event_id": "1edc517a-d049-4361-9804-e4111ce061ae", "timestamp": "2021-03-10T19:00:00Z", "event_type": "snapshot", "expiration": "2021-03-10T20:00:00Z", "instance_id": "c6123fbb-84d0-47a7-b069-ff9048caf7c7", "display_name": "c6123fbb-84d0-47a7-b069-ff9048caf7c7", "event_source": "prometheus-openshift", "measurements": [{"uom": "Cores", "value": 12.0}], "service_type": "OpenShift Cluster", "account_number": "account123"}', 'snapshot', 'prometheus-openshift', 'c6123fbb-84d0-47a7-b069-ff9048caf7c7');
INSERT INTO events (id, account_number, timestamp, data, event_type, event_source, instance_id) VALUES ('9fd69ad7-9294-4d8d-a5f4-c5ca45bdb2ec', 'account123', '2021-03-10 19:00:00.000000', '{"role":"ocp","sla": "Standard", "usage": "Development/Test", "event_id": "9fd69ad7-9294-4d8d-a5f4-c5ca45bdb2ec", "timestamp": "2021-03-10T19:00:00Z", "event_type": "snapshot", "expiration": "2021-03-10T20:00:00Z", "instance_id": "0e94471b-f9bb-4b30-8d6e-8ad6c3539392", "display_name": "0e94471b-f9bb-4b30-8d6e-8ad6c3539392", "event_source": "prometheus-openshift", "measurements": [{"uom": "Cores", "value": 12.0}], "service_type": "OpenShift Cluster", "account_number": "account123"}', 'snapshot', 'prometheus-openshift', '0e94471b-f9bb-4b30-8d6e-8ad6c3539392');
INSERT INTO events (id, account_number, timestamp, data, event_type, event_source, instance_id) VALUES ('91ef5257-6f44-4ad5-812e-ae070483e4a7', 'account123', '2021-03-10 20:00:00.000000', '{"role":"ocp","sla": "Premium", "usage": "Production", "event_id": "91ef5257-6f44-4ad5-812e-ae070483e4a7", "timestamp": "2021-03-10T20:00:00Z", "event_type": "snapshot", "expiration": "2021-03-10T21:00:00Z", "instance_id": "bdeffe2b-08c6-40ef-b2cf-e3c458a0273a", "display_name": "bdeffe2b-08c6-40ef-b2cf-e3c458a0273a", "event_source": "prometheus-openshift", "measurements": [{"uom": "Cores", "value": 12.0}], "service_type": "OpenShift Cluster", "account_number": "account123"}', 'snapshot', 'prometheus-openshift', 'bdeffe2b-08c6-40ef-b2cf-e3c458a0273a');
INSERT INTO events (id, account_number, timestamp, data, event_type, event_source, instance_id) VALUES ('a35a0f56-d46b-4c36-ad4b-e6e889e13c9f', 'account123', '2021-03-10 19:00:00.000000', '{"role":"ocp","sla": "Premium", "usage": "Production", "event_id": "a35a0f56-d46b-4c36-ad4b-e6e889e13c9f", "timestamp": "2021-03-10T19:00:00Z", "event_type": "snapshot", "expiration": "2021-03-10T20:00:00Z", "instance_id": "bdeffe2b-08c6-40ef-b2cf-e3c458a0273a", "display_name": "bdeffe2b-08c6-40ef-b2cf-e3c458a0273a", "event_source": "prometheus-openshift", "measurements": [{"uom": "Cores", "value": 12.0}], "service_type": "OpenShift Cluster", "account_number": "account123"}', 'snapshot', 'prometheus-openshift', 'bdeffe2b-08c6-40ef-b2cf-e3c458a0273a');
INSERT INTO events (id, account_number, timestamp, data, event_type, event_source, instance_id) VALUES ('0b021762-59e6-41ae-bbe6-e4ba1deaf3a1', 'account123', '2021-03-10 20:00:00.000000', '{"role":"ocp","sla": "Standard", "usage": "Development/Test", "event_id": "0b021762-59e6-41ae-bbe6-e4ba1deaf3a1", "timestamp": "2021-03-10T20:00:00Z", "event_type": "snapshot", "expiration": "2021-03-10T21:00:00Z", "instance_id": "0e94471b-f9bb-4b30-8d6e-8ad6c3539392", "display_name": "0e94471b-f9bb-4b30-8d6e-8ad6c3539392", "event_source": "prometheus-openshift", "measurements": [{"uom": "Cores", "value": 12.0}], "service_type": "OpenShift Cluster", "account_number": "account123"}', 'snapshot', 'prometheus-openshift', '0e94471b-f9bb-4b30-8d6e-8ad6c3539392');
INSERT INTO events (id, account_number, timestamp, data, event_type, event_source, instance_id) VALUES ('e223e90d-f567-4a4f-b05f-7ad9e67fe5c5', 'account123', '2021-03-10 20:00:00.000000', '{"role":"ocp","sla": "Premium", "usage": "Development/Test", "event_id": "e223e90d-f567-4a4f-b05f-7ad9e67fe5c5", "timestamp": "2021-03-10T20:00:00Z", "event_type": "snapshot", "expiration": "2021-03-10T21:00:00Z", "instance_id": "c6123fbb-84d0-47a7-b069-ff9048caf7c7", "display_name": "c6123fbb-84d0-47a7-b069-ff9048caf7c7", "event_source": "prometheus-openshift", "measurements": [{"uom": "Cores", "value": 12.0}], "service_type": "OpenShift Cluster", "account_number": "account123"}', 'snapshot', 'prometheus-openshift', 'c6123fbb-84d0-47a7-b069-ff9048caf7c7');
INSERT INTO events (id, account_number, timestamp, data, event_type, event_source, instance_id) VALUES ('922855da-29da-438f-b8c5-d2f3df7f62fd', 'account123', '2021-03-10 19:00:00.000000', '{"role":"osd","sla": "Premium", "usage": "Development/Test", "event_id": "922855da-29da-438f-b8c5-d2f3df7f62fd", "timestamp": "2021-03-10T19:00:00Z", "event_type": "snapshot", "expiration": "2021-03-10T20:00:00Z", "instance_id": "f418a78c-2099-4932-9ff2-135dabd09077", "display_name": "f418a78c-2099-4932-9ff2-135dabd09077", "event_source": "prometheus-openshift", "measurements": [{"uom": "Cores", "value": 12.0}], "service_type": "OpenShift Cluster", "account_number": "account123"}', 'snapshot', 'prometheus-openshift', 'f418a78c-2099-4932-9ff2-135dabd09077');
```

## kick off hourly tally
```sh
curl -X POST \
  'http://localhost:8080/actuator/hawtio/jolokia/' \
  -H 'Content-Type: application/json; charset=utf-8' \
  -d '{
  "type": "exec",
  "mbean": "org.candlepin.subscriptions.jmx:name=tallyJmxBean,type=TallyJmxBean",
  "operation": "tallyAccountByHourly(java.lang.String,java.lang.String,java.lang.String)",
  "arguments": [
    "account123",
    "2021-03-09T00:00:00Z",
    "2021-03-10T23:59:00Z"
  ]
}'
```

## observe that hourly snapshots are produced for the account
```sql
SELECT DISTINCT
    product_id,
    --account_number,
    granularity
  FROM tally_snapshots
 WHERE
       sla = '_ANY'
   AND usage = '_ANY'
   --AND granularity = 'HOURLY'
 GROUP BY
     product_id,
     --account_number,
     granularity
 ORDER BY
     --account_number,
     granularity ASC,
     product_id;
```
```
+---------------------------+-----------+
|product_id                 |granularity|
+---------------------------+-----------+
|OpenShift-dedicated-metrics|DAILY      |
|OpenShift-metrics          |DAILY      |
|OpenShift-dedicated-metrics|HOURLY     |
|OpenShift-metrics          |HOURLY     |
+---------------------------+-----------+
```

## kick off "regular" tally for all configured accounts

```sh
curl -X POST \
  'http://localhost:8080/actuator/hawtio/jolokia/' \
  -H 'Content-Type: application/json; charset=utf-8' \
  -d '{
  "type": "exec",
  "mbean": "org.candlepin.subscriptions.jmx:name=tallyJmxBean,type=TallyJmxBean",
  "operation": "tallyConfiguredAccounts()"
}'
```

## observe that there are no longer any RHEL/Hourly snapshots being produced
```sql
SELECT DISTINCT product_id,
    --account_number,
    granularity
FROM tally_snapshots
WHERE sla = '_ANY'
    AND usage = '_ANY'
     --AND granularity = 'HOURLY'
    --AND account_number = 'account123'
GROUP BY product_id,
    --account_number,
    granularity
ORDER BY --account_number,
    granularity ASC;
```
```
+---------------------------+-----------+
|product_id                 |granularity|
+---------------------------+-----------+
|OpenShift-dedicated-metrics|DAILY      |
|OpenShift-metrics          |DAILY      |
|RHEL                       |DAILY      |
|RHEL Desktop               |DAILY      |
|RHEL for x86               |DAILY      |
|RHEL Server                |DAILY      |
|OpenShift-dedicated-metrics|HOURLY     |
|OpenShift-metrics          |HOURLY     |
|RHEL                       |MONTHLY    |
|RHEL Desktop               |MONTHLY    |
|RHEL for x86               |MONTHLY    |
|RHEL Server                |MONTHLY    |
|RHEL                       |QUARTERLY  |
|RHEL Desktop               |QUARTERLY  |
|RHEL for x86               |QUARTERLY  |
|RHEL Server                |QUARTERLY  |
|RHEL                       |WEEKLY     |
|RHEL Desktop               |WEEKLY     |
|RHEL for x86               |WEEKLY     |
|RHEL Server                |WEEKLY     |
|RHEL                       |YEARLY     |
|RHEL Desktop               |YEARLY     |
|RHEL for x86               |YEARLY     |
|RHEL Server                |YEARLY     |
+---------------------------+-----------+

```

### CODE REVIEW

File: src/main/java/org/candlepin/subscriptions/tally/TallySnapshotController.java [hunk 1]
```diff
@@ -119,16 +120,20 @@ public void produceSnapshotsForAccounts(List<String> accounts) {
...
     public void produceHourlySnapshotsForAccount(String accountNumber, OffsetDateTime startDateTime,
         OffsetDateTime endDateTime) {
-
         try {
-            Map<OffsetDateTime, AccountUsageCalculation> accountCalcs = retryTemplate.execute(
+            Map<OffsetDateTime, AccountUsageCalculation>  accountCalcs = retryTemplate.execute(
                 context -> metricUsageCollector.collect(accountNumber, startDateTime, endDateTime));
+
+            var applicableUsageCalculations = accountCalcs.entrySet().stream()
+                .filter(TallySnapshotController::isCombiningRollupStrategySupported)
```
Comment (dmckoske) [inline src/main/java/org/candlepin/subscriptions/tally/TallySnapshotController.java:128]: I would prefer to see a strongly typed variable here instead, but it's a minor point.

**Review by dmckoske — APPROVED** [review]

File: src/main/java/org/candlepin/subscriptions/tally/TallySnapshotController.java (hunk 1 above)
Comment (lindseyburnett) [reply src/main/java/org/candlepin/subscriptions/tally/TallySnapshotController.java:128]: The variable is still strongly typed.  `var` in java lets the type be inferred by the compiler, and is just a convenience/readability feature.  It's the equivalent of below, just less noisy and easier to read in my opinion.  Easily one of the main reasons I pushed upgrading to Java 10 so much.

```java
Map<OffsetDateTime, AccountUsageCalculation> applicableUsageCalculations = accountCalcs.entrySet().stream()
    .filter(TallySnapshotController::isCombiningRollupStrategySupported)
    .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
```

@kahowell, thoughts on using `var` or not?

Comment (kahowell) [reply src/main/java/org/candlepin/subscriptions/tally/TallySnapshotController.java:128]: I find var more readable here + Intellij shows the type, so I personally prefer `var` also. Let's see what @mstead thinks as well.

Comment (kahowell) [reply src/main/java/org/candlepin/subscriptions/tally/TallySnapshotController.java:128]: Though I will say, I wouldn't bat an eyelash at explicit `Map` either.

**Comment by awood** [comment]
> I don't have instructions included for putting test data into the insights database for the HBI portion of this....

`bin/insert-mock-hosts --hbi --num-physical 10 --account acct_ --num-accounts 10 --clear`

will clear the local HBI db and insert 10 accounts with 10 physical hosts.

File: src/main/java/org/candlepin/subscriptions/tally/TallySnapshotController.java (hunk 1 above)
Comment (mstead) [reply src/main/java/org/candlepin/subscriptions/tally/TallySnapshotController.java:128]: Might be best to leave me out of this one, because I think the whole var thing is foolish for various reasons. :)

That being said, I really don't have a strong opinion one way or the other. I'm Ok with both.

**Comment by mstead** [comment]
Functions ok for me and overall it looks good. We still need to work out the fact that this PR does not remove any bad data that already exists, but prevents it from happening again. This is outside of this PR though.

File: src/main/java/org/candlepin/subscriptions/tally/TallySnapshotController.java (hunk 1 above)
Comment (mstead) [reply src/main/java/org/candlepin/subscriptions/tally/TallySnapshotController.java:128]: Perhaps @awood can weigh in. :)

**Review by mstead — APPROVED** [review]
:+1:

File: src/main/java/org/candlepin/subscriptions/tally/TallySnapshotController.java (hunk 1 above)
Comment (awood) [reply src/main/java/org/candlepin/subscriptions/tally/TallySnapshotController.java:128]: I prefer `var` for throw away stuff that's just going to be used like once or twice in close proximity to the declaration or when the type is going to be totally unambigious (e.g. `var myThing = new Thing()`).  It's also awesome for when you have to use a fully qualified class name if you have imports that clash on name (e.g. DTO classes and DB model classes).

Overall, I favor it when it is used judiciously.  The guidelines [here](https://openjdk.java.net/projects/amber/LVTIstyle.html) are intelligent and I'd be happy to use `var` in accordance with them.

File: src/main/java/org/candlepin/subscriptions/tally/TallySnapshotController.java [hunk 2]
```diff
@@ -119,16 +120,20 @@ public void produceSnapshotsForAccounts(List<String> accounts) {
     @Timed("rhsm-subscriptions.snapshots.single.hourly")
     public void produceHourlySnapshotsForAccount(String accountNumber, OffsetDateTime startDateTime,
         OffsetDateTime endDateTime) {
-
         try {
-            Map<OffsetDateTime, AccountUsageCalculation> accountCalcs = retryTemplate.execute(
+            Map<OffsetDateTime, AccountUsageCalculation>  accountCalcs = retryTemplate.execute(
```
Comment (awood) [inline src/main/java/org/candlepin/subscriptions/tally/TallySnapshotController.java:124]: Rogue space here

File: src/main/java/org/candlepin/subscriptions/tally/TallySnapshotController.java [hunk 3]
```diff
@@ -155,4 +160,13 @@ private void attemptCloudigradeEnrichment(List<String> accounts,
         }
     }
 
+    private static boolean isCombiningRollupStrategySupported(
+        Map.Entry<OffsetDateTime, AccountUsageCalculation> usageCalculations) {
+
+        var calculatedProducts = usageCalculations.getValue().getProducts();
+
+        return calculatedProducts.contains(ProductId.OPENSHIFT_METRICS.toString()) ||
```
Comment (awood) [inline src/main/java/org/candlepin/subscriptions/tally/TallySnapshotController.java:168]: Is this something we should just have in the product profile registry?  We could just add it as a new attribute (true by default) and update the `Openshift-metrics` profile.

File: build.gradle [hunk 4]
```diff
@@ -36,6 +36,7 @@ ext {
     kafka_avro_serializer_version = "6.1.0"
     avro_version = "1.10.1"
     webjars_locator_version = "0.40"
+    generatedMetamodels = "build/generated/sources/annotationProcessor/java/main/org/candlepin/subscriptions"
```
Comment (awood) [inline build.gradle:39]: Tell me more about this.  I assume it enables something cool!

**Review by awood — APPROVED** [review]
It's great the way it is.  My only thought is externalizing the `isCombiningRollupStrategySupported` predicate to the product profile.  We can always do that later though.  Let me know if you want to do that or wait and then I will merge.

File: build.gradle (hunk 4 above)
Comment (lindseyburnett) [reply build.gradle:39]: You're about to be disappointed - less cool than you'd think.

A little bit ago, I introduced Criteria, JPA Specifications, and Hibernate's Metamodel Generator to help facilitate that.  If you're using IntelliJ, the IDE is smart enough to just magically figure it out and resolve the dependencies when you have the editor open.  I started using VS Code again, which is no where near as smart as IntelliJ to figure those kind of things out, so the project/build file needs this for the VS Code gradle extension to get rid of all those "cannot resolve ______ class" warnings in the files dealing with the jpa Specifications.

Related, if VS Code didn't have a lombok extension, I'd need to add something similar to have VS Code recognize lombok annotations.  Eclipse has the same kind of issues.


TL;DR - I'm stubborn and can't give up my old IDE.

https://hibernate.org/orm/tooling/
https://docs.jboss.org/hibernate/orm/5.0/topical/html/metamodelgen/MetamodelGenerator.html

File: src/main/java/org/candlepin/subscriptions/tally/TallySnapshotController.java (hunk 3 above)
Comment (lindseyburnett) [reply src/main/java/org/candlepin/subscriptions/tally/TallySnapshotController.java:168]: I'm thinking yes.  I've actually got a whole bunch of ideas floating around in my brain about what we could/should add to the product profile registry.  I'm in mode "get this to work without bugs" right now and want to talk about the product profile registry in one of those dev backend syncs soon™

Comment (awood) [reply src/main/java/org/candlepin/subscriptions/tally/TallySnapshotController.java:168]: :+1:
